html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php

```diff
@@ -0,0 +1,151 @@
+<?php
+
+namespace Drupal\bc_dc\Controller;
+
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\Core\Entity\EntityRepositoryInterface;
+use Drupal\Core\File\FileSystemInterface;
+use Drupal\node\NodeInterface;
+use Drupal\path_alias\AliasManagerInterface;
+use PhpOffice\PhpSpreadsheet\Spreadsheet;
+use PhpOffice\PhpSpreadsheet\Worksheet\Worksheet;
+use PhpOffice\PhpSpreadsheet\Writer\Csv;
+use PhpOffice\PhpSpreadsheet\Writer\Xlsx;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\BinaryFileResponse;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+/**
+ * Create a file in csv or xlsx.
+ */
+class BcDcCreateFileController extends ControllerBase {
+
+  const SUPPORTED_EXTENSIONS = [
+    'csv',
+    'xlsx',
+  ];
+
+  /**
+   * Create a BcDcCreateFileController instance.
+   *
+   * @param \Drupal\Core\Entity\EntityRepositoryInterface $entityRepository
+   *   The entity.repository service.
+   * @param \Drupal\Core\File\FileSystemInterface $fileSystem
+   *   The file_system service.
+   * @param \Drupal\path_alias\AliasManagerInterface $pathAliasManager
+   *   The path_alias.manager service.
+   */
+  public function __construct(
+    protected EntityRepositoryInterface $entityRepository,
+    protected FileSystemInterface $fileSystem,
+    protected AliasManagerInterface $pathAliasManager,
+  ) {
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container): object {
+    return new static(
+      $container->get('entity.repository'),
+      $container->get('file_system'),
+      $container->get('path_alias.manager'),
+    );
+  }
+
+  /**
+   * Generate the download file and serves it to the browser.
+   *
+   * @param \Drupal\node\NodeInterface $node
+   *   The node.
+   * @param string $format
+   *   The format of the file to serve.
+   *
+   * @return Symfony\Component\HttpFoundation\BinaryFileResponse
+   *   The file to download.
+   */
+  public function createFile(NodeInterface $node, string $format): BinaryFileResponse {
+    // Not Found if $format is not a supported file extension.
+    if (!in_array($format, static::SUPPORTED_EXTENSIONS, TRUE)) {
+      throw new NotFoundHttpException();
+    }
+
+    $paragraph_field_items = $node->get('field_columns')->referencedEntities();
+    foreach ($paragraph_field_items as $paragraph) {
+      // Get the translation.
+      $paragraph = $this->entityRepository->getTranslationFromContext($paragraph);
+    }
+    $sheetData = [
+      [
+        'column_allowed_values',
+        'column_data_quality',
+        'column_description',
+        'column_name',
+        'column_size',
+        'column_transformations',
+        'provenance_field_name',
+        'provenance_field_number',
+        'provenance_field_question',
+        'provenance_form_name',
+        'provenance_form_number',
```

Comment:
It would be best if it did not have a hard-coded list of fields and instead included every field in the paragraph type. You can refer to the code that generates the "Download sample file" for how to do this.

Reply:
Not sure from the code in Download sample will fix hardcoding the headers

Reply:
The download sample link has code to get the names of all the fields so it does not have to have its own list.

Reply:
updated code to not hardcode headers

Reply:
The rows are still generated with a hard-coded list of fields.